kuksa_databroker/databroker-cli/src/main.rs

```diff
@@ -574,24 +574,132 @@ fn code_to_text(code: &tonic::Code) -> &str {
 }
 
 struct CliCompleter {
-    properties: HashSet<String>,
+    paths: PathPart,
 }
 
+#[derive(Debug)]
+struct PathPart {
+    rel_path: String,
+    full_path: String,
+    children: HashMap<String, PathPart>,
+}
+
+impl PathPart {
+    fn new() -> Self {
+        PathPart {
+            rel_path: "".into(),
+            full_path: "".into(),
+            children: HashMap::new(),
+        }
+    }
+}
 impl CliCompleter {
     fn new() -> CliCompleter {
         CliCompleter {
-            properties: HashSet::<String>::new(),
+            paths: PathPart::new(),
         }
     }
 
     fn from_metadata(metadata: &[proto::v1::Metadata]) -> CliCompleter {
-        let mut properties = HashSet::<String>::new();
+        let mut root = PathPart::new();
         for entry in metadata {
-            properties.insert(entry.name.to_owned());
+            let mut parent = &mut root;
+            let mut parts = entry.name.split('.');
+            loop {
+                let path = parts.next();
+                match path {
```

Comment:
I'm just curious - there's nothing wrong here, but isn't clippy complaining here that this could be a `while let`?

Reply:
No clippy warning (wouldn't pass CI if there were), but you're absolutely right that using `while let` would be better.